blog/2019-02-19-reliable-microservices-data-exchange-with-the-outbox-pattern.adoc

```diff
@@ -0,0 +1,701 @@
+Also there should be some house-keeping on the message log table;
+periodically, all events older than the consumer's current offset committed with the broker may be deleted,
+as it's ensured that such messages won't be propagated to the consumer another time.
+
+== Summary
+
+The outbox pattern is a great way for propagating data amongst different microservices.
+
+By only modifying a single resource - the source service's own database -
+it avoids any potential inconsistencies of altering multiple resources at the same time which don't share one common transactional context
+(the database and Apache Kafka).
+By writing to the database first, the source service has instant "read your own writes" semantics,
+which is important for a consistent user experience, allowing query methods invoked following to a write to instantly reflect any data changes.
+
+At the same time, the pattern enables asynchronous event propagation to other microservices.
+Apache Kafka acts as a highly scalable and reliable backbone for the messaging amongst the services.
+Given the right topic retention settings, new consumers may come up long after an event has been originally produced,
+and build up their own local state based on the event history.
+
+Putting Apache Kafka into the middle of the overall architecture also ensures a decoupling of involved services.
+If for instance single components of the solution fail or are not available for some time, e.g. during an update,
+events will simply be processed later on: after a restart,
+the Debezium connector will continue to tail the outbox table from the point where it left off before.
+Similarly, any consumer will continue to process topics from its previous offset.
+By keeping track of already successfully processed messages, duplicates can be detected and excluded from repeated handling.
+
+Naturally, such event pipeline between different services is eventually consistent,
+i.e. consumers such as the shipping service may lag a bit behind producers such as the order service.
+Usually, that's just fine, though, and can be handled in terms of the application's business logic.
+For instance there'll typically be no need to create a shipment within the very same second as an order has been placed.
+Also, end-to-end delays of the overall solution are typically low (seconds or even sub-second range),
+thanks to log-based change data capture which allows for emission of events in near-realtime.
+
+One last thing to keep in mind is that the structure of the events exposed via the outbox should be considered a part of the emitting service's API.
+I.e. when needed, their structure should be adjusted carefully and with compatibility considerations in mind.
+This is to ensure to not accidentally break any consumers when upgrading the producing service.
+At the same time, consumers should be lenient when handling messages and for instance not fail when encountering unknown attributes within received events.
```

Comment:
This last point is really important, and its the key corollary to "messaging makes you more decoupled." It may be worth making this point earlier in the blog.

Reply:
I'm touching on it earlier on already (and expanded that one a bit):

> By explicitly emitting events via records in the outbox table,
the application code can work with events are structured in a way suitable for external consumers.
> This also ensures that event consumers won't break
when for instance altering the internal domain model or the `PurchaseOrder` table.

I think it's ok that way.